internal/lib/checkpoint.go

```diff
@@ -0,0 +1,234 @@
+	config := &ContainerConfig{
+		ID:              ctr.ID(),
+		Name:            ctr.Name(),
+		RootfsImageName: ctr.ImageName(),
+	}
+
+	if _, err := metadata.WriteJSONFile(config, ctr.Dir(), metadata.ConfigDumpFile); err != nil {
+		return err
+	}
+
+	// save spec
+	jsonPath := filepath.Join(ctr.BundlePath(), "config.json")
+	g, err := generate.NewFromFile(jsonPath)
+	if err != nil {
+		return errors.Wrapf(err, "generating spec for container %q failed", ctr.ID())
+	}
+	if _, err := metadata.WriteJSONFile(g.Config, ctr.Dir(), metadata.SpecDumpFile); err != nil {
+		return errors.Wrapf(err, "generating spec for container %q failed", ctr.ID())
+	}
+
+	return nil
+}
+
+func (c *ContainerServer) exportCheckpoint(ctr *oci.Container, specgen *rspec.Spec, export string) error {
+	id := ctr.ID()
+	dest := ctr.Dir()
+	logrus.Debugf("Exporting checkpoint image of container %q to %q", id, dest)
+
+	includeFiles := []string{
+		metadata.DumpLogFile,
+		metadata.CheckpointDirectory,
+		metadata.ConfigDumpFile,
+		metadata.SpecDumpFile,
+	}
+
+	// To correctly track deleted files, let's go through the output of 'podman diff'
+	rootFsChanges, err := c.getDiff(id)
+	if err != nil {
+		return errors.Wrapf(err, "error exporting root file-system diff for %q", id)
+	}
+	mountPoint, err := c.StorageImageServer().GetStore().Mount(id, specgen.Linux.MountLabel)
+	if err != nil {
+		return errors.Wrapf(err, "Not able to get mountpoint for container %q", id)
+	}
+	addToTarFiles, err := crutils.CRCreateRootFsDiffTar(&rootFsChanges, mountPoint, dest)
+	if err != nil {
+		return err
+	}
+
+	// Put log file into checkpoint archive
+	_, err = os.Stat(specgen.Annotations[annotations.LogPath])
+	if err == nil {
+		src, err := os.Open(specgen.Annotations[annotations.LogPath])
+		if err != nil {
+			return errors.Wrapf(err, "error opening log file %q", specgen.Annotations[annotations.LogPath])
+		}
+		defer src.Close()
+		destLogPath := filepath.Join(dest, annotations.LogPath)
+		destLog, err := os.Create(destLogPath)
+		if err != nil {
+			return errors.Wrapf(err, "error opening log file %q", destLogPath)
+		}
+		defer destLog.Close()
+		_, err = io.Copy(destLog, src)
+		if err != nil {
+			return errors.Wrapf(err, "copying log file to %q failed", destLogPath)
+		}
+		addToTarFiles = append(addToTarFiles, annotations.LogPath)
+	}
+
+	includeFiles = append(includeFiles, addToTarFiles...)
+
+	input, err := archive.TarWithOptions(ctr.Dir(), &archive.TarOptions{
+		// This should be configurable via api.proti
+		Compression:      archive.Uncompressed,
+		IncludeSourceDir: true,
+		IncludeFiles:     includeFiles,
+	})
+	if err != nil {
+		return errors.Wrapf(err, "error reading checkpoint directory %q", id)
+	}
+
+	outFile, err := os.Create(export)
+	if err != nil {
+		return errors.Wrapf(err, "error creating checkpoint export file %q", export)
+	}
+	defer outFile.Close()
+
+	// The resulting tar archive should not readable by everyone as it contains
```

Comment:
*should not be readable or writable by anyone except its owner...

Reply:
changed